File: src/com/castlefrog/agl/domains/hex/HexAction.java
package com.castlefrog.agl.domains.hex;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import com.castlefrog.agl.Action;

/**
 * Immutable object that represents a hex action.
 */
public final class HexAction implements Action, Serializable {
    private static final long serialVersionUID = 1L;

    private static List<List<HexAction>> actions_ = new ArrayList<List<HexAction>>();

    private final byte x_;
    private final byte y_;

    private HexAction(int x,
                      int y) {
        x_ = (byte) x;
        y_ = (byte) y;
    }

    /**
     * Returns a hex action.
     * @param x x-coord
     * @param y y-coord
     * @return
     *      action corrisponding to x and y coord
     */
    public static HexAction valueOf(int x, int y) {
        if (x >= actions_.size() || y >= actions_.get(0).size()) {
            generateActions(Math.max(x, y) + 1);
        }
        return actions_.get(x).get(y);
    }

    private static void generateActions(int size) {
        for (int i = 0; i < size; i += 1) {
            if (i >= actions_.size()) {
                actions_.add(new ArrayList<HexAction>());
            }
            for (int j = actions_.get(i).size(); j < size; j += 1) {
                actions_.get(i).add(new HexAction(i, j));
            }
        }
    }

    public HexAction copy() {
        return this;
    }

    public int getX() {
        return x_;
    }

    public int getY() {
        return y_;
    }

    @Override
    public int hashCode() {
        return x_ + 31 * y_;
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof HexAction)) {
            return false;
        }
        HexAction action = (HexAction) object;
        return x_ == action.getX() && y_ == action.getY();
    }

    @Override
    public String toString() {
        return "(" + ((char) (0x41 + x_)) + y_ + ")";
    }
}

File: api/src/main/java/com/castlefrog/agl/domains/hex/HexAction.kt
package com.castlefrog.agl.domains.hex

import java.io.Serializable
import java.util.ArrayList

import com.castlefrog.agl.Action

/**
 * Immutable object that represents a hex action.
 */
data class HexAction private constructor(val x: Byte, val y: Byte) : Action, Serializable {

    override fun copy(): HexAction {
        return this
    }

    override fun toString(): String {
        return "" + (0x41 + x).toChar() + y
    }

    companion object {
        private val actions = ArrayList<MutableList<HexAction>>()

        /**
         * Returns a hex action.
         * @param x x-coordinate
         * @param y y-coordinate
         * @return action corresponding to x and y coordinate
         */
        fun valueOf(x: Int, y: Int): HexAction {
            if (x >= actions.size || y >= actions[0].size) {
                generateActions(Math.max(x, y) + 1)
            }
            return actions[x][y]
        }

        private fun generateActions(size: Int) {
            for (i in 0..size - 1) {
                if (i >= actions.size) {
                    actions.add(ArrayList<HexAction>())
                }
                for (j in actions[i].size..size - 1) {
                    actions[i].add(HexAction(i.toByte(), j.toByte()))
                }
            }
        }
    }
}
